// @pluralsight/ps-design-system-datepicker/dist/esm/react/datepicker.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { canUseDOM, generateId } from '@pluralsight/ps-design-system-util';
import { format } from 'date-fns';
import { useDayzed } from 'dayzed';
import React from 'react';
import { Calendar } from './calendar';
import { CalendarDay } from './calendar-day';
import { TextInputField } from './text-input-field';
import { handleDateSelectChange } from './utils';
export var DatePicker = function (_a) {
    var disabled = _a.disabled, error = _a.error, label = _a.label, prefix = _a.prefix, renderContainer = _a.renderContainer, renderTag = _a.renderTag, size = _a.size, subLabel = _a.subLabel, suffix = _a.suffix, onSelect = _a.onSelect, valueFromProps = _a.value, props = __rest(_a, ["disabled", "error", "label", "prefix", "renderContainer", "renderTag", "size", "subLabel", "suffix", "onSelect", "value"]);
    var _b = React.useState(valueFromProps), selected = _b[0], setSelected = _b[1];
    var ref = React.useRef();
    React.useEffect(function () { return setSelected(valueFromProps); }, [valueFromProps]);
    var _c = React.useState(false), open = _c[0], setOpen = _c[1];
    var _d = React.useState(''), value = _d[0], setValue = _d[1];
    var focusInput = function () {
        var _a;
        var wrapper = ref.current;
        wrapper && ((_a = wrapper.querySelector('input')) === null || _a === void 0 ? void 0 : _a.focus());
    };
    var dateFormat = 'MM/dd/yyyy';
    var onDateSelected = function (dateObj, evt) {
        focusInput();
        var nextSelected = dateObj.date;
        onSelect && onSelect(evt, dateObj);
        setSelected(nextSelected);
        setValue(format(nextSelected, dateFormat));
        setOpen(false);
    };
    var handleIconClick = function (evt) {
        open && focusInput();
        setOpen(!open);
    };
    var handleTextfieldKeyDown = function (evt) {
        var key = evt.key.toLowerCase();
        [' ', 'enter'].includes(key) && setOpen(true);
    };
    var handleEscapeKeyDown = function (evt) {
        var key = evt.key.toLowerCase();
        if (key === 'escape') {
            focusInput();
            setOpen(false);
        }
    };
    var handleFocus = function (evt) {
        if (open) {
            focusInput();
            setOpen(false);
        }
    };
    React.useEffect(function () {
        if (!canUseDOM())
            return function () { };
        var handleClickOutsideMenu = function (evt) {
            if (evt.target instanceof HTMLElement) {
                // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
                if (ref.current.contains(evt.target))
                    return;
                setOpen(false);
            }
        };
        document.addEventListener('click', handleClickOutsideMenu, {
            capture: true
        });
        return function () {
            return document.removeEventListener('click', handleClickOutsideMenu, {
                capture: true
            });
        };
    }, [setOpen]);
    var labelId = generateId('psds-datepicker-text-input__label-');
    var inputId = generateId('psds-datepicker-text-input__input-');
    var handleChange = function (evt) {
        var nextValue = evt.target.value;
        setValue(nextValue);
        handleDateSelectChange({
            selected: selected,
            setSelected: setSelected,
            value: nextValue,
            dateFormat: dateFormat
        });
    };
    return (React.createElement("div", __assign({ style: { display: 'inline-block', position: 'relative' } }, props, { ref: ref }),
        React.createElement(TextInputField, { disabled: disabled, error: error, label: label, onChange: handleChange, onClick: handleIconClick, onKeyDown: handleTextfieldKeyDown, onFocus: handleFocus, placeholder: "mm/dd/yyyy", prefix: prefix, renderContainer: renderContainer, renderTag: renderTag, size: size, subLabel: subLabel, suffix: suffix, value: value, labelId: labelId, inputId: inputId }),
        React.createElement("br", null),
        open && (React.createElement(CalendarWrapper, { selected: selected, onDateSelected: onDateSelected, labelId: labelId, handleEscapeKeyDown: handleEscapeKeyDown }))));
};
var CalendarWrapper = function (_a) {
    var _b = _a.selected, selected = _b === void 0 ? new Date() : _b, onDateSelected = _a.onDateSelected, labelId = _a.labelId, handleEscapeKeyDown = _a.handleEscapeKeyDown;
    return (React.createElement(Calendar, __assign({}, useDayzed({
        date: selected,
        selected: selected,
        onDateSelected: onDateSelected
    }), { autofocus: false, trapped: true, returnFocus: false, "aria-modal": "true", "aria-labelledby": labelId, "aria-live": "polite", style: { position: 'absolute', zIndex: 1, marginTop: 4 }, onKeyDown: handleEscapeKeyDown, selected: selected }), function (props) { return React.createElement(CalendarDay, __assign({}, props)); }));
};
